src/scf/newton.jl

```diff
@@ -0,0 +1,420 @@
+#
+# Newton algorithm consist of iterating over density matrices like
+#       P = P - δP
+# where δP solves
+#       (Ω+K)δP = [P,[P,H(P)]] (residual)
+# where (Ω+K) if the Jacobian of the direct minimization iterations. It is
+# defined as a super operator from the space tangent to the constraints
+# manifold at point P_∞, the solution of our problem.
+#   - K is the Hessian of the energy on the manifold;
+#   - Ω represents the influence of the curvature of the manifold :
+#         ΩδP = -[P_∞,[H(P_∞),δP]].
+#     In practice, we dont have access to P_∞ so we just use the current P.
+#     Ω is also -χ0^(-1) where χ0 is the occupation function.
+#
+# Here, in the orbital framework, an element on the tangent space at φ can be written as
+#       δP = Σ |φi><δφi| + hc
+# where the δφi are of size Nb and are all orthogonal to the φj, 1 <= j <= N.
+# Therefore we store them in the same kind of array than φ, with
+# δφ[ik][:,i] = δφi for each k-point.
+# In this framework :
+#   - computing Ωδφ can be done analitically with the formula
+#         Ωδφi = H*δφi - Σj <δφj|H|δφi> δφj - Σj λji δφj
+#     where λij = <φi|H|φj>;
+#   - computing Kδφ can be done using the kernel function of DFTK after
+#     computing the associated δρ;
+#   - to perform Newton iterations, we solve (Ω+K)δφ = Hφ-λφ and then perform
+#     φ = φ - δφ, with proper orthonormalization.
+#
+# For further details :
+# Eric Cancès, Gaspard Kemlin, Antoine Levitt. Convergence analysis of
+# direct minimization and self-consistent iterations. SIAM Journal of Matrix
+# Analysis and Applications, 42(1):243–274 (2021)
+
+import KrylovKit: ArnoldiIterator, Orthogonalizer, OrthonormalBasis, KrylovDefaults, orthogonalize!
+using KrylovKit
+
+#
+# Tools
+#
+
+"""
+    compute_residual(basis::PlaneWaveBasis, φ, occ)
+
+Compute the residual associated to a set of planewave φ, that is to say
+H(φ)*φ - λ*φ where λ is the set of rayleigh coefficients associated to the φ.
+"""
+function compute_residual(basis::PlaneWaveBasis, φ, occ)
+
+    # necessary quantities
+    Nk = length(basis.kpoints)
+    ρ = compute_density(basis, φ, occ)
+    energies, H = energy_hamiltonian(basis, φ, occ; ρ=ρ)
+
+    # compute residual
+    res = similar(φ)
+    for ik = 1:Nk
+        φk = φ[ik]
+        N = size(φk, 2)
+        Hk = H.blocks[ik]
+        # eigenvalues as rayleigh coefficients
+        egvalk = φk'*(Hk*φk)
+        # compute residual at given kpoint as H(φ)φ - λφ
+        res[ik] = Hk*φk - φk*egvalk
+    end
+    res
+end
+
+# To project onto the space tangent to φ, we project δφ onto the orthogonal of φ
+function proj_tangent(δφ, φ; tol_test=1e-12)
+
+    Nk1 = size(δφ,1)
+    Nk2 = size(φ,1)
+    @assert Nk1 == Nk2
+    Nk = Nk1
+
+    Πδφ = similar(δφ)
+
+    for ik = 1:Nk
+        φk = φ[ik]
+        δφk = δφ[ik]
+        Πδφk = deepcopy(δφk)
+
+        N1 = size(δφk,2)
+        N2 = size(φk,2)
+        @assert N1 == N2
+        N = N1
```

Comment:
I think a better pattern (less code) is:
```suggestion
        @assert size(δφk,2) == size(φk,2)
        N = size(φk,2)
```